#!/usr/bin/python3

#----------------------------------------------------------------------------#
#
# Microbiome Insights | Snakemake pipeline for microbial characterization
#
#----------------------------------------------------------------------------#


# Module Imports
#----------------------------------------------------------------------------#
import os, sys, re, itertools, argparse, psutil, copy, yaml, configparser

import pandas as pd
from pathlib import Path
from datetime import datetime
now = datetime.now().strftime("%d/%m/%Y %H:%M:%S")

# Configuration 
#----------------------------------------------------------------------------#

#Load config from M5P yaml
if config:
    #M5P CONFIG
    working_dir = config["working_dir"]
    fastq_metagenomics = config["fastq_metagenomics"]
    fastq_metatranscriptomics = config["fastq_metatranscriptomics"]
    genome = config["genome"]
    database_dir  = config["database_dir"]
    THREADS = config["threads"]
    merged_reads  = config["merged_reads"]
    merged_prefix = config["merged_prefix"]
    metadata_path = config["metadata_path"]
    bin_all = config["bin_all"]
else:
    raise ImportError("Error: no config parameters were provided.")

# ---- Import Snakemake rules
include: "workflows/rules/common.smk"
include: "workflows/rules/custom_atlas.smk"
include: "workflows/rules/custom_dram.smk"
include: "workflows/rules/clean_after_atlas.smk"
include: "workflows/rules/run_bakta.smk"
include: "workflows/rules/run_grist.smk"
include: "workflows/rules/custom_praxis.smk"
include: "workflows/rules/run_referenceseeker.smk"
include: "workflows/rules/clean_after_praxis.smk"


# ---- Rule all: run all targets
rule all:
    input: COLLECT_ALL_INPUT()
